File: Analysis/satd.py
"""
Analysis/satd.py — Self-Admitted Technical Debt (SATD) Scanner (v8.0)
======================================================================

Scans Python source files for SATD markers in comments and docstrings
(TODO, FIXME, HACK, XXX, DEBT, WORKAROUND, BUG, NOQA) and classifies
them by type with estimated remediation hours.

Usage::

    from Analysis.satd import SATDScanner, SATDItem

    scanner = SATDScanner()
    items = scanner.scan_directory(Path("/some/project"))
    total_hours = sum(i.hours for i in items)

Research basis:
    DebtViz (arXiv 2023), Transformer-based SATD classifiers (arXiv 2024-2025).
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import List, Optional

# ── Marker patterns → SATD category ──────────────────────────────────────────

_MARKERS: list[tuple[re.Pattern, str]] = [
    (re.compile(r"\b(FIXME)\b|\b(BUGFIX|BUG)\s*:", re.IGNORECASE), "defect"),
    (re.compile(r"\b(HACK|WORKAROUND|KLUDGE|KLUGE)\b", re.IGNORECASE), "design"),
    (re.compile(r"\b(TODO|TO-DO|TO DO)\b", re.IGNORECASE), "design"),
    (re.compile(r"\b(DEBT|TECH-DEBT|TECHNICAL DEBT)\b", re.IGNORECASE), "debt"),
    (re.compile(r"\b(XXX)\b"), "defect"),
    (re.compile(r"\b(NOQA|TYPE:\s*IGNORE)\b", re.IGNORECASE), "test"),
    (re.compile(r"\b(REMOVEME|DEPRECATED|REMOVE ME)\b", re.IGNORECASE), "design"),
    (re.compile(r"\b(OPTIMIZE|PERF|PERFORMANCE)\b", re.IGNORECASE), "design"),
    (re.compile(r"\b(SECURITY|SEC ISSUE)\b", re.IGNORECASE), "defect"),
    (re.compile(r"\b(DOCME|DOCUMENT|UNDOCUMENTED)\b", re.IGNORECASE), "documentation"),
]

# Estimated remediation hours per SATD category (conservative)
_HOURS: dict[str, float] = {
    "defect": 1.0,
    "design": 2.0,
    "debt": 3.0,
    "test": 0.5,
    "documentation": 0.25,
}

# Comment patterns (single-line and inline)
_COMMENT_RE = re.compile(r"#\s*(.*)")
_DOCSTRING_RE = re.compile(r'"""(.*?)"""|\'\'\'(.*?)\'\'\'', re.DOTALL)

# Skip non-source directories
_SKIP_DIRS = frozenset(
    {
        "__pycache__",
        ".git",
        ".venv",
        "venv",
        "node_modules",
        "dist",
        "build",
        ".mypy_cache",
        ".pytest_cache",
        "_rustified",
        "_training_ground",
        "_mothership",
        "_verify_crate",
    }
)


@dataclass
class SATDItem:
    """A single self-admitted technical debt occurrence."""

    file: str
    line: int
    category: str  # defect / design / debt / test / documentation
    marker: str  # The matched keyword (e.g. "TODO")
    text: str  # Full comment text (stripped)
    hours: float  # Estimated remediation hours
    severity: str = "warning"

    @property
    def short_text(self) -> str:
        """Return text capped at 120 chars."""
        return self.text[:120] + ("…" if len(self.text) > 120 else "")


@dataclass
class SATDSummary:
    """Aggregate result of a SATD scan."""

    items: List[SATDItem] = field(default_factory=list)

    @property
    def total(self) -> int:
        return len(self.items)

    @property
    def total_hours(self) -> float:
        return round(sum(i.hours for i in self.items), 1)

    @property
    def by_category(self) -> dict[str, list[SATDItem]]:
        cats: dict[str, list[SATDItem]] = {}
        for item in self.items:
            cats.setdefault(item.category, []).append(item)
        return cats

    @property
    def by_file(self) -> dict[str, list[SATDItem]]:
        files: dict[str, list[SATDItem]] = {}
        for item in self.items:
            files.setdefault(item.file, []).append(item)
        return files

    @property
    def top_files(self) -> list[tuple[str, int, float]]:
        """Return list of (file, count, hours) sorted by count desc."""
        result = []
        for f, items in self.by_file.items():
            result.append((f, len(items), sum(i.hours for i in items)))
        return sorted(result, key=lambda x: x[1], reverse=True)

    def as_dict(self) -> dict:
        return {
            "total": self.total,
            "total_hours": self.total_hours,
            "by_category": {cat: len(items) for cat, items in self.by_category.items()},
            "top_files": [
                {"file": f, "count": c, "hours": round(h, 1)}
                for f, c, h in self.top_files[:10]
            ],
            "items": [
                {
                    "file": i.file,
                    "line": i.line,
                    "category": i.category,
                    "marker": i.marker,
                    "text": i.short_text,
                    "hours": i.hours,
                }
                for i in self.items
            ],
        }


class SATDScanner:
    """Scan a directory tree for self-admitted technical debt comments."""

    def __init__(self, extensions: tuple[str, ...] = (".py",)):
        self.extensions = extensions

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def scan_directory(self, root: Path) -> SATDSummary:
        """Recursively scan *root* for SATD and return a SATDSummary."""
        summary = SATDSummary()
        for path in self._walk(root):
            summary.items.extend(self._scan_file(path, root))
        return summary

    def scan_files(self, files: list[Path], root: Optional[Path] = None) -> SATDSummary:
        """Scan an explicit list of files."""
        summary = SATDSummary()
        for path in files:
            summary.items.extend(self._scan_file(path, root or path.parent))
        return summary

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _walk(self, root: Path):
        for path in root.rglob("*"):
            if any(part in _SKIP_DIRS for part in path.parts):
                continue
            if path.suffix in self.extensions and path.is_file():
                yield path

    def _scan_file(self, path: Path, root: Path) -> list[SATDItem]:
        items: list[SATDItem] = []
        try:
            text = path.read_text(encoding="utf-8", errors="replace")
        except OSError:
            return items

        rel = str(path.relative_to(root)) if root else str(path)

        for lineno, line in enumerate(text.splitlines(), start=1):
            comment_match = _COMMENT_RE.search(line)
            if comment_match:
                comment_text = comment_match.group(1).strip()
                item = self._classify(rel, lineno, comment_text)
                if item:
                    items.append(item)

        return items

    @staticmethod
    def _classify(file: str, line: int, text: str) -> Optional[SATDItem]:
        """Return a SATDItem if *text* contains a marker, else None."""
        for pattern, category in _MARKERS:
            m = pattern.search(text)
            if m:
                marker = m.group(0).upper().strip()
                hours = _HOURS.get(category, 1.0)
                return SATDItem(
                    file=file,
                    line=line,
                    category=category,
                    marker=marker,
                    text=text,
                    hours=hours,
                )
        return None

File: scanner/src/analyzers/satd.rs
//! SATD Scanner — Self-Admitted Technical Debt analysis.
//! Rust port of services/satd_scanner.py.

use regex::Regex;
use std::collections::HashMap;
use walkdir::WalkDir;

use crate::constants::{SKIP_DIRS, TEXT_EXTS};

struct SatdMarker {
    pattern: Regex,
    category: &'static str,
    hours: f64,
}

fn satd_markers() -> Vec<SatdMarker> {
    vec![
        SatdMarker { pattern: Regex::new(r"(?i)\b(FIXME)\b").unwrap(), category: "defect", hours: 1.0 },
        SatdMarker { pattern: Regex::new(r"(?i)\b(BUG|BUGFIX)\b").unwrap(), category: "defect", hours: 1.0 },
        SatdMarker { pattern: Regex::new(r"(?i)\b(XXX)\b").unwrap(), category: "defect", hours: 1.0 },
        SatdMarker { pattern: Regex::new(r"(?i)\b(SECURITY)\b").unwrap(), category: "defect", hours: 1.0 },
        SatdMarker { pattern: Regex::new(r"(?i)\b(HACK)\b").unwrap(), category: "design", hours: 2.0 },
        SatdMarker { pattern: Regex::new(r"(?i)\b(WORKAROUND)\b").unwrap(), category: "design", hours: 2.0 },
        SatdMarker { pattern: Regex::new(r"(?i)\b(KLUDGE)\b").unwrap(), category: "design", hours: 2.0 },
        SatdMarker { pattern: Regex::new(r"(?i)\b(TODO)\b").unwrap(), category: "design", hours: 2.0 },
        SatdMarker { pattern: Regex::new(r"(?i)\b(OPTIMIZE|PERF)\b").unwrap(), category: "design", hours: 2.0 },
        SatdMarker { pattern: Regex::new(r"(?i)\b(TECH.?DEBT|DEBT)\b").unwrap(), category: "debt", hours: 3.0 },
        SatdMarker { pattern: Regex::new(r"(?i)\b(NOQA|type:\s*ignore)\b").unwrap(), category: "test", hours: 0.5 },
        SatdMarker { pattern: Regex::new(r"(?i)\b(DOCME|DOCUMENT|UNDOCUMENTED)\b").unwrap(), category: "documentation", hours: 0.25 },
    ]
}

/// Forward-slash path normaliser.
fn fwd(p: &str) -> String {
    p.replace('\\', "/")
}

/// Scan for Self-Admitted Technical Debt markers (TODO, FIXME, HACK, etc.).
pub fn scan_satd(directory: &str) -> serde_json::Value {
    let markers = satd_markers();
    let comment_re = Regex::new(r"#\s*(.*)").unwrap();

    let mut items: Vec<serde_json::Value> = Vec::new();
    let mut by_category: HashMap<String, Vec<serde_json::Value>> = HashMap::new();
    let mut total_hours: f64 = 0.0;

    for entry in WalkDir::new(directory)
        .into_iter()
        .filter_entry(|e| {
            if e.file_type().is_dir() {
                let name = e.file_name().to_str().unwrap_or("");
                !SKIP_DIRS.contains(&name) && !name.starts_with('.')
            } else {
                true
            }
        })
        .filter_map(|e| e.ok())
    {
        if !entry.file_type().is_file() {
            continue;
        }
        let path = entry.path();
        let ext = path.extension().and_then(|e| e.to_str()).map(|e| format!(".{}", e));
        let ext_str = ext.as_deref().unwrap_or("");
        if !TEXT_EXTS.contains(&ext_str) {
            continue;
        }

        let content = match std::fs::read_to_string(path) {
            Ok(c) => c,
            Err(_) => continue,
        };
        let rel = path.strip_prefix(directory).unwrap_or(path)
            .to_string_lossy().to_string();

        for (lineno, line) in content.lines().enumerate() {
            let lineno = lineno + 1;
            for marker in &markers {
                if let Some(m) = marker.pattern.find(line) {
                    let text = if let Some(cm) = comment_re.captures(line) {
                        cm.get(1).map(|x| x.as_str().trim()).unwrap_or(line.trim())
                    } else {
                        line.trim()
                    };
                    let text = if text.len() > 200 { &text[..200] } else { text };
                    let matched_marker = m.as_str().to_uppercase();
                    let item = serde_json::json!({
                        "file": fwd(&rel),
                        "line": lineno,
                        "category": marker.category,
                        "marker": matched_marker,
                        "text": text,
                        "hours": marker.hours,
                    });
                    total_hours += marker.hours;
                    by_category.entry(marker.category.to_string()).or_default().push(item.clone());
                    items.push(item);
                    break; // first match wins per line
                }
            }
        }
    }

    serde_json::json!({
        "total_items": items.len(),
        "total_hours": (total_hours * 10.0).round() / 10.0,
        "items": items,
        "by_category": by_category,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_scan_satd_finds_markers() {
        let tmp = tempfile::tempdir().unwrap();
        let proj = tmp.path().join("project");
        std::fs::create_dir_all(&proj).unwrap();
        std::fs::write(proj.join("main.py"), "# TODO: fix this\n# FIXME: urgent\ncode = 1\n").unwrap();
        let result = scan_satd(proj.to_str().unwrap());
        assert!(result["total_items"].as_u64().unwrap() >= 2);
    }

    #[test]
    fn test_scan_satd_empty_dir() {
        let tmp = tempfile::tempdir().unwrap();
        let proj = tmp.path().join("project");
        std::fs::create_dir_all(&proj).unwrap();
        let result = scan_satd(proj.to_str().unwrap());
        assert_eq!(result["total_items"].as_u64().unwrap(), 0);
    }
}
